flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskManagerMetricGroup.java

```diff
@@ -81,61 +79,33 @@ public String taskManagerId() {
     //  job groups
     // ------------------------------------------------------------------------
 
-    public TaskMetricGroup addTaskForJob(
-            final JobID jobId,
-            final String jobName,
-            final JobVertexID jobVertexId,
-            final ExecutionAttemptID executionAttemptId,
-            final String taskName,
-            final int subtaskIndex,
-            final int attemptNumber) {
+    public TaskManagerJobMetricGroup getOrCreateTaskManagerJobMetricGroup(
+            JobID jobId, String jobName) {
         Preconditions.checkNotNull(jobId);
-
         String resolvedJobName = jobName == null || jobName.isEmpty() ? jobId.toString() : jobName;
-
-        // we cannot strictly lock both our map modification and the job group modification
-        // because it might lead to a deadlock
-        while (true) {
-            // get or create a jobs metric group
-            TaskManagerJobMetricGroup currentJobGroup;
-            synchronized (this) {
-                currentJobGroup = jobs.get(jobId);
-
-                if (currentJobGroup == null || currentJobGroup.isClosed()) {
-                    currentJobGroup =
-                            new TaskManagerJobMetricGroup(registry, this, jobId, resolvedJobName);
-                    jobs.put(jobId, currentJobGroup);
-                }
-            }
-
-            // try to add another task. this may fail if we found a pre-existing job metrics
-            // group and it is closed concurrently
-            TaskMetricGroup taskGroup =
-                    currentJobGroup.addTask(
-                            jobVertexId, executionAttemptId, taskName, subtaskIndex, attemptNumber);
-
-            if (taskGroup != null) {
-                // successfully added the next task
-                return taskGroup;
+        TaskManagerJobMetricGroup jobGroup;
+        synchronized (this) { // synchronization isn't strictly necessary as of FLINK-24864
+            jobGroup = jobs.get(jobId);
+            if (jobGroup == null) {
+                jobGroup = new TaskManagerJobMetricGroup(registry, this, jobId, resolvedJobName);
+                jobs.put(jobId, jobGroup);
             }
-
-            // else fall through the loop
         }
+        return jobGroup;
     }
 
-    public void removeJobMetricsGroup(JobID jobId, TaskManagerJobMetricGroup group) {
-        if (jobId == null || group == null || !group.isClosed()) {
-            return;
-        }
-
-        synchronized (this) {
-            // optimistically remove the currently contained group, and check later if it was
-            // correct
-            TaskManagerJobMetricGroup containedGroup = jobs.remove(jobId);
+    public TaskManagerJobMetricGroup getJobMetricsGroup(JobID jobId) {
```

Comment:
Can we mark this as `@VisibleForTesting`?

Reply:
I'm planning to use it in `TaskExecutor` in #17757.

Reply:
But...couldn't you just pass `jobGroup`?

Reply:
Yes, you're right :)